### PR TITLE
CWL v1.0.1 support

### PR DESCRIPTION
**Pull request by tetron**
Work in progress, not ready to merge.

Update cwltoil to complete CWL v1.0.1

### CODE REVIEW

**Comment by tetron** [comment]
Now passing CWL conformance tests 100% locally but blocked on a couple of other branches that need to merge first:

CWL conformance test fix: https://github.com/common-workflow-language/common-workflow-language/pull/496

cwltool updates: https://github.com/common-workflow-language/cwltool/pull/516

File: src/toil/cwl/cwltoil.py [hunk 1]
```diff
@@ -323,6 +387,16 @@ def makeJob(tool, jobobj, **kwargs):
         wfjob.addFollowOn(followOn)
         return (wfjob, followOn)
     else:
+        # get_requirement
+        resourceReq, _ = tool.get_requirement("ResourceRequirement")
+        for req in ("coresMin", "coresMax", "ramMin", "ramMax",
+                     "tmpdirMin", "tmpdirMax", "outdirMin", "outdirMax"):
+            r = resourceReq.get(req)
+            if isinstance(r, string_types) and "$(" in r or "${" in r:
```
Comment (mr-c) [inline src/toil/cwl/cwltoil.py]: From the logs (I forget if you have permission to see them, my apologies for the spam if you do):
```
Test failed: /mnt/ephemeral/workspace/toil-pull-requests/venv/bin/cwltoil --outdir=/mnt/ephemeral/tmp/tmpFSMzsM --quiet v1.0/bwa-mem-tool.cwl v1.0/bwa-mem-job.json
General test of command line generation
Returned non-zero
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/venv/bin/cwltoil", line 9, in <module>
    load_entry_point('toil', 'console_scripts', 'cwltoil')()
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/cwl/cwltoil.py", line 810, in main
    tmpdir=os.path.realpath(outdir))
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/cwl/cwltoil.py", line 395, in makeJob
    if isinstance(r, string_types) and "$(" in r or "${" in r:
TypeError: argument of type 'int' is not iterable
```
Which is odd and doesn't make sense to me :-/

File: src/toil/cwl/cwltoil.py [hunk 2]
```diff
@@ -323,6 +387,16 @@ def makeJob(tool, jobobj, **kwargs):
         wfjob.addFollowOn(followOn)
         return (wfjob, followOn)
     else:
+        # get_requirement
+        resourceReq, _ = tool.get_requirement("ResourceRequirement")
+        for req in ("coresMin", "coresMax", "ramMin", "ramMax",
+                     "tmpdirMin", "tmpdirMax", "outdirMin", "outdirMax"):
+            r = resourceReq.get(req)
```
Comment (mr-c) [inline src/toil/cwl/cwltoil.py]: From the next conformance test:
```

Test failed: /mnt/ephemeral/workspace/toil-pull-requests/venv/bin/cwltoil --outdir=/mnt/ephemeral/tmp/tmpQQ9YJF --quiet v1.0/binding-test.cwl v1.0/bwa-mem-job.json
Test nested prefixes with arrays
Returned non-zero
Traceback (most recent call last):
  File "/mnt/ephemeral/workspace/toil-pull-requests/venv/bin/cwltoil", line 9, in <module>
    load_entry_point('toil', 'console_scripts', 'cwltoil')()
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/cwl/cwltoil.py", line 810, in main
    tmpdir=os.path.realpath(outdir))
  File "/mnt/ephemeral/workspace/toil-pull-requests/src/toil/cwl/cwltoil.py", line 394, in makeJob
    r = resourceReq.get(req)
AttributeError: 'NoneType' object has no attribute 'get'
```

**Comment by tetron** [comment]
Thanks for looking at it @mr-c!  I think I see what I did wrong, just silly mistakes.

**Comment by mr-c** [comment]
Jenkins, test this please

**Comment by mr-c** [comment]
Oh yeah, the tests pass! Now to try it out at EBI with LSF :-)

**Comment by stefanoberri** [comment]
Hi.

I hope not to interfere with the PR process, but I made a PR (#1824) to this branch that contains a couple of unit tests around reading from and writing to directories.

The tests pass on my system, however I have noticed something "strange". One of the key/values of the generated json contain the following

`'http://commonwl.org/cwltool#generation': 0,`

when generating a directory. I am not sure what that means and the url does not appear to be valid.

**Comment by tetron** [comment]
@stefanoberri thanks for the PR with additional testing.  

The "generation" field is related to a cwltool-specific extension: https://github.com/common-workflow-language/cwltool/issues/382 it shouldn't be leaking out like that, I'll see about adding a pass to remove it from the final output.

**Comment by stefanoberri** [comment]
Thank you for implementing the directory support!
It looks like that when jenkins run the tests it does not work ;(

**Comment by tetron** [comment]
Looks like it failed for an unrelated reason: `Post https://quay.io/v2/ucsc_cgl/toil-it/blobs/uploads/: net/http: TLS handshake timeout` ping @ejacox

File: src/toil/cwl/cwltoil.py [hunk 3]
```diff
@@ -266,18 +330,19 @@ def __init__(self, tool, cwljob, **kwargs):
                                      memory=(req["ram"]*1024*1024),
                                      disk=((req["tmpdirSize"]*1024*1024) + (req["outdirSize"]*1024*1024)),
                                      unitName=unitName)
-        #super(CWLJob, self).__init__()
```
Comment (mr-c) [inline src/toil/cwl/cwltoil.py]: @tetron Was this removal on purpose?

File: src/toil/cwl/cwltoil.py [hunk 4]
```diff
@@ -323,6 +388,17 @@ def makeJob(tool, jobobj, **kwargs):
...
     else:
+        # get_requirement
+        resourceReq, _ = tool.get_requirement("ResourceRequirement")
+        if resourceReq:
+            for req in ("coresMin", "coresMax", "ramMin", "ramMax",
+                         "tmpdirMin", "tmpdirMax", "outdirMin", "outdirMax"):
+                r = resourceReq.get(req)
+                if isinstance(r, string_types) and ("$(" in r or "${" in r):
+                    # Found a dynamic resource requirement so use a job wrapper.
+                    job = CWLJobWrapper(tool, jobobj, **kwargs)
```
Comment (mr-c) [inline src/toil/cwl/cwltoil.py]: What will the resource requirements be for the wrapper around the real job?

**Comment by jessebrennan** [comment]
> Looks like it failed for an unrelated reason: Post https://quay.io/v2/ucsc_cgl/toil-it/blobs/uploads/: net/http: TLS handshake timeout

This is being addressed in #1822.

**Comment by mr-c** [comment]
@tetron Sorry about that extra commit, I meant to put it in another branch as a PR to this one

**Comment by mr-c** [comment]
@tetron I'm still getting a File Literal related error: https://github.com/BD2KGenomics/toil/issues/1836#issue-250580172

File: src/toil/cwl/cwltoil.py (hunk 3 above)
Comment (tetron) [inline src/toil/cwl/cwltoil.py]: Yes, because the previous line is already calling super(CWLJob, self).

File: src/toil/cwl/cwltoil.py (hunk 4 above)
Comment (tetron) [inline src/toil/cwl/cwltoil.py]: The wrapper has minimal resources coded in the wrapper's super(CWLJobWrapper, self).__init__ method (but you figured that out already).

**Comment by tetron** [comment]
@ejacox I've added code documentation and confirmed that Toil in this branch works with bcbio.  I think it's ready for a look by someone on the core Toil team.

**Comment by chapmanb** [comment]
+1 This branch looks great to me based on local tests with the bcbio test CWL. Thanks for all this great work Peter, Michael, Edwin and all. Looking forward to testing this at scale on some bigger workflows.

**Comment by ejacox** [comment]
Thank you for the reviews @mr-c and @chapmanb

File: docs/architecture.rst [hunk 5]
```diff
@@ -136,3 +136,63 @@ disk-based storage systems as compared to the SSD systems we tested this on.
...
+local files and object storage.
+
+The 'location' field of File references are updated to reflect the import token
+returned by the Toil file store.
+
+For directory inputs, the directory listing is stored in Directory object.
+Each individual files is imported into Toil file store.
+
+The initial workflow Job is created from the toplevel CWL document, and
+
```
Comment (ejacox) [inline docs/architecture.rst]: Is there more to add here?

**Comment by DailyDreaming** [comment]
On AWS, for the GA4GH-DREAM challenges, cwltoil is giving a similar error for bcbio, gdc dna-seq, and biowardrobe ( 
[example_error.txt](https://github.com/BD2KGenomics/toil/files/1232463/example_error.txt)
 ) with the same cwltool "failure to evaluate the expression" (using cwltoil).

Also, #1796 using the maxCores option I assume can be addressed by adding this option to the "hints" section recently created in the last commit for pushing the --defaultCores, etc options.

**Comment by chapmanb** [comment]
For the Javascript expression issues with cwltool, what node.js do you have installed on your machine? It looks like something is problematic with the setup but Peter might have might idea exactly what.

For `--maxCores`, we need more of a global metric in Toil rather than task specific changes in the CWL support. We can already specify correctly how many cores/CPU a particular task uses but would like to be able to say something like: Run these 100 tasks each using 8 cores and 32Gb of memory, but only allocate up to 64 total cores at a time. Then Toil would run these in batches for 8 at a time, cycling through the 100. I believe this works cleanly on single machines and AWS but not sure about on batch systems. Thanks much for looking at this.

File: src/toil/cwl/cwltoil.py (hunk 3 above)
Comment (mr-c) [inline src/toil/cwl/cwltoil.py]: D'oh, I see that now, sorry for the noise

**Comment by DailyDreaming** [comment]
@chapmanb , it was node.js.  Thank you, I should have seen that.

I appreciate laying out the description.  I was misunderstanding the problem.  I'll try testing on a batch system with the bcbio files.

File: docs/architecture.rst (hunk 5 above)
Comment (tetron) [inline docs/architecture.rst]: Sorry, writing fail.  Fixed in ea77add4081d0d8359f0674b1ccc27d0877f252f

**Comment by tetron** [comment]
@DailyDreaming did you get it working?  I agree with @chapmanb that it is related to the use of nodejs to evaluate expressions but I'm not really sure what is going wrong there.

**Comment by DailyDreaming** [comment]
@tetron I did not, but it is giving me a new error now at least: 
[new_error.txt](https://github.com/BD2KGenomics/toil/files/1235628/new_error.txt)

**Comment by chapmanb** [comment]
It looks like things are running now but there is something problematic with workflow. It's complaining about not finding any outputs but it looks like the command line passes "False" so doesn't actually run the rmdup command. I don't really know much about that workflow, it might be worth asking the authors if they have any pointers.

**Comment by DailyDreaming** [comment]
@chapmanb I'll do that.  Thanks for the suggestion.
